Model/Service/Product/Attribute/CachingAttributeService.php

```diff
@@ -0,0 +1,165 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Service\Product\Attribute;
+
+use Magento\Catalog\Model\Product;
+use Magento\Eav\Model\Entity\Attribute\AbstractAttribute;
+use Magento\Store\Api\Data\StoreInterface;
+use Nosto\Tagging\Helper\Data as NostoHelperData;
+use Nosto\Tagging\Logger\Logger as NostoLogger;
+
+class CachingAttributeService extends AbstractAttributeService
+{
+    /** @var array */
+    private $productAttributeCache = [];
+
+    /** @var AttributeServiceInterface */
+    private $attributeService;
+
+    /** @var int */
+    private $maxCachedProducts;
+
+    /**
+     * CachingAttributeService constructor.
+     * @param AttributeServiceInterface $attributeService
+     * @param $maxCachedProducts
+     */
+    public function __construct(
+        NostoHelperData $nostoHelperData,
+        NostoLogger $logger,
+        AttributeServiceInterface $attributeService,
+        AttributeProviderInterface $attributeProvider,
+        $maxCachedProducts
+    ) {
+        parent::__construct($nostoHelperData, $logger, $attributeProvider);
+        $this->attributeService = $attributeService;
+        $this->maxCachedProducts = $maxCachedProducts;
+    }
+
+    /**
+     * Saves a single attribute and value into the cache.
+     *
+     * @param Product $product
+     * @param StoreInterface $store
+     * @param $attributeCode
+     * @param $value
+     */
+    private function saveAttributeToCache(Product $product, StoreInterface $store, $attributeCode, $value)
+    {
+        $storeId = $store->getId();
+        $productId = $product->getId();
+        if (!isset($this->productAttributeCache[$storeId])) {
+            $this->productAttributeCache[$storeId] = [];
+        }
+        if (!isset($this->productAttributeCache[$storeId][$productId])) {
+            $this->productAttributeCache[$storeId][$productId] = [];
+        }
+        $this->productAttributeCache[$storeId][$productId][$attributeCode] = $value;
+        $this->sliceCache($store);
+    }
+
+    /**
+     * Returns the value of given product, store and attribute code from the cache
+     * if it exists.
+     *
+     * @param Product $product
+     * @param StoreInterface $store
+     * @param $attributeCode
+     * @return |null
```

Comment:
Maybe you mean `string|null`?
```suggestion
     * @return string|null
```